libglusterfs/src/dict.c

```diff
@@ -1999,28 +1978,19 @@ _dict_modify_flag(dict_t *this, char *key, int flag, int op)
             else
                 BIT_CLEAR((unsigned char *)(data->data), flag);
 
-            if (this->free_pair.key) { /* the free pair is in use */
-                pair = mem_get0(THIS->ctx->dict_pair_pool);
-                if (!pair) {
-                    gf_smsg("dict", GF_LOG_ERROR, ENOMEM, LG_MSG_NO_MEMORY,
-                            "dict pair", NULL);
-                    ret = -ENOMEM;
-                    goto err;
-                }
-            } else { /* use the free pair */
-                pair = &this->free_pair;
-            }
-
-            pair->key = (char *)GF_MALLOC(strlen(key) + 1, gf_common_mt_char);
-            if (!pair->key) {
+            keylen = strlen(key) + 1;  // including terminating NULL char
+            pair = GF_MALLOC(sizeof(data_pair_t) + keylen,
+                             gf_common_mt_data_pair_t);
+            if (caa_unlikely(!pair)) {
                 gf_smsg("dict", GF_LOG_ERROR, ENOMEM, LG_MSG_NO_MEMORY,
                         "dict pair", NULL);
                 ret = -ENOMEM;
                 goto err;
             }
-            strcpy(pair->key, key);
+
             pair->value = data_ref(data);
-            this->totkvlen += (strlen(key) + 1 + data->len);
+            strcpy(pair->key, key);
```

Comment:
Knowing the key length we can use this:
```suggestion
            memcpy(pair->key, key, keylen);
```